File: app/src/main/java/com/worldventures/dreamtrips/modules/tripsimages/service/command/SendProgressAnalyticsIfNeed.java
package com.worldventures.dreamtrips.modules.tripsimages.service.command;

import com.worldventures.dreamtrips.core.utils.tracksystem.BaseAnalyticsAction;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public abstract class SendProgressAnalyticsIfNeed<E extends BaseAnalyticsAction> extends Command<Integer> {

   private int expectedAnalyticStep;
   private final long currentProgress;
   private final long totalLength;

   public SendProgressAnalyticsIfNeed(int expectedAnalyticStep, long currentProgress, long totalLength) {
      this.expectedAnalyticStep = expectedAnalyticStep;
      this.currentProgress = currentProgress;
      this.totalLength = totalLength;
   }

   @Override
   protected void run(CommandCallback<Integer> callback) throws Throwable {
      int percent = (int) ((currentProgress * 100) / totalLength);
      int currentStep = percent / 25;
      E action = chooseAnalyticAction(currentStep, expectedAnalyticStep);
      if (action != null) {
         sendAnalyticAction(action);
         callback.onSuccess(expectedAnalyticStep + 1);
      } else
         callback.onSuccess(expectedAnalyticStep);
   }

   protected abstract E chooseAnalyticAction(int currentStep, int expectedAnalyticStep);

   protected abstract void sendAnalyticAction(E action);
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/service/analytics/SendProgressAnalyticsIfNeed.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.service.analytics

import com.worldventures.core.service.analytics.BaseAnalyticsAction

import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction

@CommandAction
abstract class SendProgressAnalyticsIfNeed<E : BaseAnalyticsAction>(private val expectedAnalyticStep: Int,
                                                                    private val currentProgress: Long,
                                                                    private val totalLength: Long) : Command<Int>() {

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Int>) {
      val percent = (currentProgress * 100 / totalLength).toInt()
      val currentStep = percent / STEP_PERCENT
      val action = chooseAnalyticAction(currentStep, expectedAnalyticStep)
      if (action != null) {
         sendAnalyticAction(action)
         callback.onSuccess(expectedAnalyticStep + 1)
      } else {
         callback.onSuccess(expectedAnalyticStep)
      }
   }

   protected abstract fun chooseAnalyticAction(currentStep: Int, expectedAnalyticStep: Int): E?

   protected abstract fun sendAnalyticAction(action: E)

   companion object {
      const val STEP_PERCENT = 25
   }
}
